test/blocklib/core/unit-test/tag_monitors.hpp

```diff
@@ -41,31 +41,31 @@ struct TagSource : public node<TagSource<T, UseProcessOne>> {
     OUT<T>             out;
     std::vector<tag_t> tags{};
     std::size_t        next_tag{ 0 };
-    std::uint64_t      n_samples_max = 1024;
-    std::uint64_t      n_samples_produced{ 0 };
```

Comment:
FWIW, `n_samples_produced = -1` is not *ill-formed*. You can make it fail to compile using `-Werror`. But assigning `-1` to an unsigned integer has a well-defined and portable meaning in standard C++.
IMNSHO you should use a different type to actually make assignment of signed integers ill-formed.

Semantically, with `unsigned` you're using a type where `x > 0 and y > 0 and (x + y < x or x - y > x or ...)` can be `true`. The default type should be `signed` if you want the usual semantics of a value in $\mathbb{N}$ or $\mathbb{Z}$.

And like I wrote in the commit message:
1. The return value of available_samples is signed.
2. The tag index is signed.
3. Subtraction of two n_samples_* values should never wrap but rather
yield a negative value.

Reply:
I appreciate your comprehensive explanation regarding the choice between signed and unsigned integer types.

However, I still believe that `n_samples_produced` should naturally be an unsigned type. This is based on the mathematical definition of *natural numbers*, which include '0' and all positive integers. The notion of `n_samples_produced = -1` would be the start of an 'ill-formed' representation or, humorously, a 'bad dad joke'; you need to add one to have zero samples produced. This implicitly indicates a flawed API design.

While I understand that the current architecture necessitates `available_samples(..) -> int` and `DONE`, I must insist that these feel like a patchwork solution, the change from unsigned to signed integer was not necessary or relevant to the proposed PR, and overcomplicates the design for everyone involved. Our aim should be to progress towards an architecture that does not require an intricate interplay between two functions to produce a correct outcome.

This particular design choice has already led to misunderstandings and costly mistakes for at least three developers, resulting in several days of lost productivity. This issue requires extensive documentation and explanation, which is a strong indicator of a less-than-optimal design.

We should consider alternatives to improve the intuitiveness of our codebase while still satisfying the functional requirements. This will help us avoid similar issues in the future and create a more accessible and maintainable codebase.

Thank you for your understanding and the continued discussion on improving the API.

Reply:
> However, I still believe that `n_samples_produced` should naturally be an unsigned type. This is based on the mathematical definition of _natural numbers_, which include '0' and all positive integers.

`unsigned` integers in C++ are modulo rings ($\mathbb{Z}/n\mathbb{Z}$) not natural numbers ($\mathbb{N}_0$). Also [ES.106: Don’t try to avoid negative values by using `unsigned`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es106-dont-try-to-avoid-negative-values-by-using-unsigned)

> The notion of `n_samples_produced = -1` would be the start of an 'ill-formed' representation or, humorously, a 'bad dad joke'; you need to add one to have zero samples produced. This implicitly indicates a flawed API design.

```c++
unsigned n_samples_produced = 0u - 1u; // worse bad dad joke? because:
assert(n_samples_produced >= 0);
n_samples_produced += 1;
// zero samples produced
```

> While I understand that the current architecture necessitates `available_samples(..) -> int` and `DONE`, I must insist that these feel like a patchwork solution,

> the change from unsigned to signed integer was not necessary or relevant to the proposed PR, and overcomplicates the design for everyone involved. Our aim should be to progress towards an architecture that does not require an intricate interplay between two functions to produce a correct outcome.

This is why I made it a separate commit; easy to drop. The reason I made the change was because I had to modify the `available_samples` function and the unsigned type made it harder to reason about the arithmetics (for me). If only I find it more challenging to reason about modulo rings ($\mathbb{Z}/n\mathbb{Z}$) rather than $\mathbb{Z}$ then I'll take the extra time and think harder... FWIW I'm not alone: [ES.102: Use signed types for arithmetic](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es102-use-signed-types-for-arithmetic)

> This particular design choice has already led to misunderstandings and costly mistakes for at least three developers, resulting in several days of lost productivity. This issue requires extensive documentation and explanation, which is a strong indicator of a less-than-optimal design.

Can you be more specific? The choice for available_samples to return a signed integer? Or special values for available_samples?